File: data-structure-and-algorithm/oc/minimum-path-sum.m
// 64. Minimum Path Sum
// Difficulty: Medium

// Given a m x n grid filled with non-negative numbers,
// find a path from top left to bottom right which minimizes the sum of all numbers along its path.

// Note: You can only move either down or right at any point in time.

// Time : O(m * n)
// Space: O(n)

#import <Foundation/Foundation.h>
int minPathSum(NSArray* grid) {
  const NSInteger m = [grid count];
  const NSInteger n = [grid[0] count];
  int f[n];
  for (int i = 0; i < n; i++) {
    f[i] = INT_MAX;
  }
  f[0] = 0;
  for (int i = 0; i < m; i++) {
    f[0] += [grid[i][0] intValue];
    for (int j = 1; j < n; j++) {
      f[j] = MIN(f[j - 1], f[j]) + [grid[i][j] intValue];
    }
  }
  return f[n - 1];
}

File: data-structure-and-algorithm/javascript/leetcode/minimum-path-sum.js
/*
Given a m x n grid filled with non-negative numbers, find a path from top left to bottom right, which minimizes the sum of all numbers along its path.

Note: You can only move either down or right at any point in time.

 

Example 1:


Input: grid = [[1,3,1],[1,5,1],[4,2,1]]
Output: 7
Explanation: Because the path 1 → 3 → 1 → 1 → 1 minimizes the sum.
Example 2:

Input: grid = [[1,2,3],[4,5,6]]
Output: 12
 

Constraints:

m == grid.length
n == grid[i].length
1 <= m, n <= 200
0 <= grid[i][j] <= 100
*/

function minPathSum(grid) {
    const [rows, cols] = [grid.length,  grid[0].length];
    const dp = Array(cols).fill(Infinity);
    dp[0] = 0;
    for (let i = 0; i < rows; ++i) {
        dp[0] += grid[i][0];
        for (let j = 1; j < cols; ++j) {
            dp[j] = Math.min(dp[j - 1], dp[j]) + grid[i][j];
        }
    }
    return dp[cols - 1];
}